perlock-core/src/main/java/name/mitterdorfer/perlock/impl/WatchServicePathWatcher.java

```diff
@@ -75,11 +83,13 @@ private void performRegistration() throws IOException {
      * Process all events for keys queued to the watcher
      */
     @Override
-    public void start() throws IOException {
+    public PathWatcher start() throws IOException {
         prepareWatcher();
         performRegistration();
         running = true;
+        lifecycleListener.onStart(this);
```

Comment:
I think this line is problematic for multiple reasons:

1. We have no control whatsoever over the lifecycle listener, meaning that it could throw an exception due to a bug in the implementation of the lifecycle listener and we don't detect it (by default the thread will just terminate and that's it but you'll never see the exception). That's why the original code used the `Silently.run()` idiom.
2. By providing `this` to the lifecycle listener we're leaking an implementation detail to the listener. I think this can be mitigated by creating a wrapper lifecycle listener that will then replace the sender with the outer class (i.e. `RunnablePathWatcherAdapter`). I hope you know how I mean it. :)